src/NuGet.Core/NuGet.Logging/CollectorLogger.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace NuGet.Logging
+{
+    public class CollectorLogger : ICollectorLogger
```

Comment:
I suggest just making this an ErrorCollectorLogger and skipping the dictionary lookup for other types. The perf difference is probably negligible but I don't see any scenarios where we would want to hold all the verbose messages in memory.